src/optimizer.cpp

```diff
@@ -0,0 +1,45 @@
+#include <Eigen/Core>
+
+#include <deepworks/optimizer.hpp>
+#include <util/assert.hpp>
+
+namespace deepworks {
+namespace optimizer {
+
+using ConstMatrix = Eigen::Map<const Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>;
+using Matrix = Eigen::Map<Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>;
+
+SGD::SGD(Parameters& params, float lr) : m_params(params), m_lr(lr) {}
+
+void SGD::step() {
+    for (auto& param: m_params) {
+        if (param.is_trainable()) {
+            auto weight = param.data();
+            const auto grad = param.grad();
+
+            const auto& shape = weight.shape();
+
+            DeepWorks_Assert(grad.shape() == shape);
+            DeepWorks_Assert(shape.size() == 2 || shape.size() == 1);
```

Comment:
For convolution `shape.size() = 4`. I think we don't need this check.

Reply:
ty, fixed it
i live in 2d world